src/init/features/apphosting/githubConnections.ts

```diff
@@ -0,0 +1,437 @@
+  const cloneUri = await promptOnce({
+    type: "autocomplete",
+    name: "cloneUri",
+    message: "Which of the following repositories would you like to deploy?",
+    source: (_: any, input = ""): Promise<(inquirer.DistinctChoice | inquirer.Separator)[]> => {
+      return new Promise((resolve) =>
+        resolve([
+          new inquirer.Separator(),
+          {
+            name: "Missing a repo? Select this option to configure your GitHub connection settings",
+            value: ADD_CONN_CHOICE,
+          },
+          new inquirer.Separator(),
+          ...fuzzy
+            .filter(input, cloneUris, {
+              extract: (uri) => extractRepoSlugFromUri(uri) || "",
+            })
+            .map((result) => {
+              return {
+                name: extractRepoSlugFromUri(result.original) || "",
+                value: result.original,
+              };
+            }),
+        ]),
+      );
+    },
+  });
+  return { cloneUri, connection: cloneUriToConnection[cloneUri] };
+}
+
+/**
+ * Exported for unit testing
+ */
+export async function ensureSecretManagerAdminGrant(projectId: string): Promise<void> {
+  const projectNumber = await getProjectNumber({ projectId });
+  const dcsaEmail = devConnect.serviceAgentEmail(projectNumber);
+
+  // will return false even if the service account does not exist in the project
+  const alreadyGranted = await rm.serviceAccountHasRoles(
+    projectId,
+    dcsaEmail,
+    ["roles/secretmanager.admin"],
+    true,
+  );
+  if (alreadyGranted) {
+    utils.logBullet("secret manager admin role already granted");
+    return;
+  }
+
+  utils.logBullet(
+    "To create a new GitHub connection, Secret Manager Admin role (roles/secretmanager.admin) is required on the Developer Connect Service Agent.",
+  );
+  const grant = await promptOnce({
+    type: "confirm",
+    message: "Grant the required role to the Developer Connect Service Agent?",
+  });
+  if (!grant) {
+    utils.logBullet(
+      "You, or your project administrator, should run the following command to grant the required role:\n\n" +
+        "You, or your project adminstrator, can run the following command to grant the required role manually:\n\n" +
+        `\tgcloud projects add-iam-policy-binding ${projectId} \\\n` +
+        `\t  --member="serviceAccount:${dcsaEmail} \\\n` +
+        `\t  --role="roles/secretmanager.admin\n`,
+    );
+    throw new FirebaseError("Insufficient IAM permissions to create a new connection to GitHub");
+  }
+
+  try {
+    await rm.addServiceAccountToRoles(
+      projectId,
+      dcsaEmail,
+      ["roles/secretmanager.admin"],
+      /* skipAccountLookup= */ true,
+    );
+  } catch (e: any) {
+    // if the dev connect P4SA doesn't exist in the project, generate one
+    if (e?.code === 400 || e?.status === 400) {
+      await devConnect.generateP4SA(projectNumber);
+      await rm.addServiceAccountToRoles(
```

Comment:
> You're right in that calling GenerateServiceIdentity multiple times after the P4SA already exists will always return a done operation (eg from your testing), but this may not be the case for a new onboarding project that doesn't already have a p4sa.

No I meant on the first call. I've tested it a couple times, creating a new project from scratch. On first call to GenerateServiceIdentity we get an operation with the name `finished.DONE_OPERATION ` with `done = true`. And polling with that operation name results in a 400 error. I can add the polling logic and skip polling if the returned operation is already done.